File: leetcode/#108.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    TreeNode* bst(vector<int>& nums, int start, int end) {
        if (start >= end) {
            return NULL;
        }

        int mid = (end + start) / 2;

        TreeNode* root = new TreeNode(nums[mid]);

        root->right = bst(nums, mid + 1, end);
        root->left = bst(nums, start, mid);

        return root;
    }


    TreeNode* sortedArrayToBST(vector<int>& nums) {

        return bst(nums, 0, nums.size());

    }
};

File: leetcode/#108.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def create_bst(self, l, r, nums):
        if l >= r:
            return None
        
        mid = (l + r) // 2
        
        node = TreeNode(nums[mid])
        
        node.left = self.create_bst(l, mid, nums)
        node.right = self.create_bst(mid + 1, r, nums)
        
        return node
        
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        
        return self.create_bst(0, len(nums), nums)